README.md
# Advent Of Code 2017 (Clojure)

An attempt to re-catch myself with Clojure, and hopefully stick with it,
if you find something that you think can be improved you are more than
welcomed to make a contribution.

## Project structure

Input files will be stored at ```resources/dayXX.txt```
and solutions in ```src/advent_of_code_2017/dayXX.clj```.
The file ```src/advent_of_code_2017/common.clj``` will have
some code that is used by multiple solutions, and will also
exist a test file in ```test/advent_of_code_2017/core_test.clj```

## Usage

Inside the project folder, the command:

``` text
lein run
```

will show the solution for every puzzle so far implemented:

``` text
$ lein run

  ===============================================
 | Day | Part | Solution                         |
 |===============================================|
 |  1  |  A   | 1216                             |
 |  1  |  B   | 1072                             |
 |  2  |  A   | 58975                            |
 |  2  |  B   | 308                              |
 |  3  |  A   | 552                              |
 |  3  |  B   | 330785                           |
 |  4  |  A   | 337                              |
 |  4  |  B   | 231                              |
 |  5  |  A   | 373160                           |
 |  5  |  B   | 26395586                         |
 |  6  |  A   | 12841                            |
 |  6  |  B   | 8038                             |
 |  7  |  A   | cyrupz                           |
 |  7  |  B   | 193                              |
 |  8  |  A   | 4066                             |
 |  8  |  B   | 4829                             |
 |  9  |  A   | 11089                            |
 |  9  |  B   | 5288                             |
 | 10  |  A   | 15990                            |
 | 10  |  B   | 90adb097dd55dea8305c900372258ac6 |
 | 11  |  A   | 818                              |
 | 11  |  B   | 1596                             |
 | 12  |  A   | 175                              |
 | 12  |  B   | 213                              |
 | 13  |  A   | 1476                             |
 | 13  |  B   | 3937334                          |
 | 14  |  A   | 8250                             |
 | 14  |  B   | 1113                             |
 | 15  |  A   | 650                              |
 | 15  |  B   | 336                              |
 | 16  |  A   | ociedpjbmfnkhlga                 |
 | 16  |  B   | gnflbkojhicpmead                 |
 | 17  |  A   | 1912                             |
 | 17  |  B   | 21066990                         |
 | 18  |  A   | 8600                             |
 | 18  |  B   | 7239                             |
 | 19  |  A   | MKXOIHZNBL                       |
 | 19  |  B   | 17872                            |
 | 20  |  A   | 119                              |
 | 20  |  B   | 471                              |
 | 21  |  A   | 152                              |
 | 21  |  B   | 1956174                          |
 | 22  |  A   | 5538                             |
 | 22  |  B   | 2511090                          |
 | 23  |  A   | 5929                             |
 | 23  |  B   | 907                              |
 | 24  |  A   | 1868                             |
 | 24  |  B   | 1841                             |
 | 25  |  A   | 4225                             |
 | 25  |  B   | Happy AoC !                      |
  ===============================================
```

## [LICENSE](https://github.com/Average-user/adventofcode-clj-2017/tree/master/LICENSE)

``` text
MIT License

Copyright (c) 2017 Lucas Polymeris

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-12-21
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-12-21
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/advent-of-code-2017/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/advent-of-code-2017/compare/0.1.0...0.1.1


src/advent_of_code_2017/core.clj
(ns advent-of-code-2017.core
  (:require [advent-of-code-2017.day01 :as day01]
            [advent-of-code-2017.day02 :as day02]
            [advent-of-code-2017.day03 :as day03]
            [advent-of-code-2017.day04 :as day04]
            [advent-of-code-2017.day05 :as day05]
            [advent-of-code-2017.day06 :as day06]
            [advent-of-code-2017.day07 :as day07]
            [advent-of-code-2017.day08 :as day08]
            [advent-of-code-2017.day09 :as day09]
            [advent-of-code-2017.day10 :as day10]
            [advent-of-code-2017.day11 :as day11]
            [advent-of-code-2017.day12 :as day12]
            [advent-of-code-2017.day13 :as day13]
            [advent-of-code-2017.day14 :as day14]
            [advent-of-code-2017.day15 :as day15]
            [advent-of-code-2017.day16 :as day16]
            [advent-of-code-2017.day17 :as day17]
            [advent-of-code-2017.day18 :as day18]
            [advent-of-code-2017.day19 :as day19]
            [advent-of-code-2017.day20 :as day20]
            [advent-of-code-2017.day21 :as day21]
            [advent-of-code-2017.day22 :as day22] 
            [advent-of-code-2017.day23 :as day23]
            [advent-of-code-2017.day24 :as day24]
            [advent-of-code-2017.day25 :as day25]))

(defn my-format [x]
  (str x (apply str (take (- 32 (count (str x))) (repeat " "))) " |"))
  

(defn -main []
  (do
    (println "\n  ===============================================")
    (println " | Day | Part | Solution                         |")
    (println " |===============================================|")
    (println (str " |  1  |  A   | " (my-format (day01/part-1))))
    (println (str " |  1  |  B   | " (my-format (day01/part-2))))
    (println (str " |  2  |  A   | " (my-format (day02/part-1))))
    (println (str " |  2  |  B   | " (my-format (day02/part-2))))
    (println (str " |  3  |  A   | " (my-format (day03/part-1))))
    (println (str " |  3  |  B   | " (my-format (day03/part-2))))
    (println (str " |  4  |  A   | " (my-format (day04/part-1))))
    (println (str " |  4  |  B   | " (my-format (day04/part-2))))
    (println (str " |  5  |  A   | " (my-format (day05/part-1))))
    (println (str " |  5  |  B   | " (my-format (day05/part-2))))
    (println (str " |  6  |  A   | " (my-format (day06/part-1))))
    (println (str " |  6  |  B   | " (my-format (day06/part-2))))
    (println (str " |  7  |  A   | " (my-format (day07/part-1))))
    (println (str " |  7  |  B   | " (my-format (day07/part-2))))
    (println (str " |  8  |  A   | " (my-format (day08/part-1))))
    (println (str " |  8  |  B   | " (my-format (day08/part-2))))
    (println (str " |  9  |  A   | " (my-format (day09/part-1))))
    (println (str " |  9  |  B   | " (my-format (day09/part-2))))
    (println (str " | 10  |  A   | " (my-format (day10/part-1))))
    (println (str " | 10  |  B   | " (my-format (day10/part-2))))
    (println (str " | 11  |  A   | " (my-format (day11/part-1))))
    (println (str " | 11  |  B   | " (my-format (day11/part-2))))
    (println (str " | 12  |  A   | " (my-format (day12/part-1))))
    (println (str " | 12  |  B   | " (my-format (day12/part-2))))
    (println (str " | 13  |  A   | " (my-format (day13/part-1))))
    (println (str " | 13  |  B   | " (my-format (day13/part-2))))
    (println (str " | 14  |  A   | " (my-format (day14/part-1))))
    (println (str " | 14  |  B   | " (my-format (day14/part-2))))
    (println (str " | 15  |  A   | " (my-format (day15/part-1))))
    (println (str " | 15  |  B   | " (my-format (day15/part-2))))
    (println (str " | 16  |  A   | " (my-format (day16/part-1))))
    (println (str " | 16  |  B   | " (my-format (day16/part-2))))
    (println (str " | 17  |  A   | " (my-format (day17/part-1))))
    (println (str " | 17  |  B   | " (my-format (day17/part-2))))
    (println (str " | 18  |  A   | " (my-format (day18/part-1))))
    (println (str " | 18  |  B   | " (my-format (day18/part-2))))
    (println (str " | 19  |  A   | " (my-format (day19/part-1))))
    (println (str " | 19  |  B   | " (my-format (day19/part-2))))
    (println (str " | 20  |  A   | " (my-format (day20/part-1))))
    (println (str " | 20  |  B   | " (my-format (day20/part-2))))
    (println (str " | 21  |  A   | " (my-format (day21/part-1))))
    (println (str " | 21  |  B   | " (my-format (day21/part-2))))
    (println (str " | 22  |  A   | " (my-format (day22/part-1))))
    (println (str " | 22  |  B   | " (my-format (day22/part-2))))
    (println (str " | 23  |  A   | " (my-format (day23/part-1))))
    (println (str " | 23  |  B   | " (my-format (day23/part-2))))
    (println (str " | 24  |  A   | " (my-format (day24/part-1))))
    (println (str " | 24  |  B   | " (my-format (day24/part-2))))
    (println (str " | 25  |  A   | " (my-format (day25/part-1))))
    (println (str " | 25  |  B   | " (my-format "Happy AoC !")))
    (println "  ===============================================")))
    


